### PR TITLE
lora : add support for non-llama models

### LINKED ISSUE #3319: [Lora] Cannot convert falcon lora to GGUF.
# Expected Behavior

convert-lora-to-gguf.py converts a HF lora adapter bin to GGUF format.

# Current Behavior

Spits out the error  Error: unrecognized tensor 

# Environment and Context

I think at least some stuff has to be added to:

```
HF_SUBLAYER_TO_GGML = {
    "self_attn.q_proj": "attn_q",
    "self_attn.k_proj": "attn_k",
    "self_attn.v_proj": "attn_v",
    "self_attn.o_proj": "attn_output",
    "mlp.gate_proj": "ffn_gate",
    "mlp.down_proj": "ffn_down",
    "mlp.up_proj": "ffn_up",
    "input_layernorm": "attn_norm",
    "post_attention_layernorm": "ffn_norm",
}
```

And maybe some other llama specific things changed down the line but I have no idea of what the names are supposed to be or where to find them.

Now that merging adapters is possible, this would let all of the falcon 180b tunes work without downloading 100GB+ each.

### LINKED ISSUE #4499: Failed to apply lora adapter after fine-tuning a llama-2-13B-chat with ./finetune
Hi everybody,

I am trying to fine-tune a llama-2-13B-chat model and I think I did everything correctly but I still cannot apply my lora.

What I did was:

1) I converted the llama2 weights into hf format using this: (https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/convert_llama_weights_to_hf.py)

python convert_llama_weights_to_hf.py --input_dir ../models/llama-2-13b-chat --output_dir ../models/llama-2-13b-chat/llama-2-13b-chat-hf --model_size 13B

2) Then I ran the convert.py and converted it in fp32

python convert.py ../models/llama-2-13b-chat/llama-2-13b-chat-hf --outtype f32 --outfile ../models/llama-2-13b-chat/llama-2-13b-chat-hf-f32.bin

3) then I quantised it (I am using q5_k_m and not q4_0)

./quantize ../models/llama-2-13b-chat/llama-2-13b-chat-hf-f32.bin ../models/llama-2-13b-chat/llama-2-13b-chat-hf-quantized_q5_k_m.bin q5_k_m 

At this point I test the models with ./main and they work perfectly.

4) So I created a dataset trying to orientate myself in many contradictory answers about the format and at the end I opted for the one that seemed the most used:

------------------------------------------------------------------------------------------------------------------------------

> &lt;s&gt;[INST] &lt;&lt;SYS&gt;&gt;
> In the context of physics
> &lt;&lt;/SYS&gt;&gt;
> 
> What is quantum entanglement? [/INST] Quantum entanglement is the phenomenon that occurs when a group of  particles are generated, interact, or share in such a way that the quantum state of each particle of the group cannot be described independently of the state of the others, including when the particles are separated by a large distance.

------------------------------------------------------------------------------------------------------------------------------

(I didn't put any &lt;/s&gt; at the end because for some reason the loss became nan after less than 10 iterations)

5) I started the fine-tuning and left it for almost 12h for completing 1 epoch

./finetune --model-base ../models/llama-2-13b-chat/llama-2-13b-chat-hf-quantized_q5_k_m.bin --train-data ../datasets/FineTune/train_llamacpp.txt --threads 26 --sample-start "&lt;s&gt;" --ctx 512 -ngl 32

6) Then I tested the model

./main -i -m ../models/llama-2-13b-chat/llama-2-13b-chat-hf-quantized_q5_k_m.bin --lora-base ../models/llama-2-13b-chat/llama-2-13b-chat-hf-f32.bin --lora ../models/llama-2-13b-chat/ggml-lora-LATEST-f32.gguf --color -p "What is entanglement in physics?"

I always get as last lines of the logs

.....
llama_apply_lora_from_file_internal: unsupported tensor dimension 1
llama_init_from_gpt_params: error: failed to apply lora adapter
ggml_metal_free: deallocating
main: error: unable to load model

I am running everything on an M1 Max with 64GB of Ram and 1 GPU with 32 cores

What can be the problem? I tried already different things but no success, that's why I'm writing here...

Thank you for any help

Luca

### PR DESCRIPTION
**Pull request by slaren**
Fixes #3319

### CODE REVIEW

**Comment by Ph0rk0z** [comment]
Thanks! I was able to convert the falcon lora after specifying falcon as the type. Now I'm going to see if merging them works from xades fork. So far the merges themselves were successful. Will load them and try inference next.

**Comment by Ph0rk0z** [comment]
All my merged lora generate gibberish unfortunately. I am surprised it is crashing, maybe something else is wrong? I will try using https://github.com/xaedes/llama.cpp/tree/finetune-lora/examples/export-lora to merge some more llama lora, I only tested on one 70b before but it didn't have this problem. That model worked normally.

**Comment by slaren** [comment]
If you are having any issues, please let me know how to reproduce it, otherwise there is not much I can do about it. As with llama, applying a lora to a quantized model may produce poor results. I think that the `export-lora` example should work as is, but that's beyond the scope of this PR, so if you have any issues with that, it would be better to report it to @xaedes.

**Review by ggerganov — APPROVED** [review]

**Comment by Ph0rk0z** [comment]
To reproduce I converted:

https://huggingface.co/quantumaikr/falcon-180B-wizard_alpaca_dolly_orca-LoRA

https://huggingface.co/SkunkworksAI/falcon_180B_Platypus_QLORA_v2

then I merged them to falcon Q3KM (https://huggingface.co/TheBloke/Falcon-180B-Chat-GGUF/tree/main) and got garbled results.

Same process for llama-70b, including merging multiple lora at full scale is yielding excellent results and even minimal loss of perplexity when I run the test.

**Comment by BarfingLemurs** [comment]
@Ph0rk0z maybe try with the base model too.

**Comment by Ph0rk0z** [comment]
I need to d/l the base model. For llama I am merging non base and it's working. Perhaps I should also d/l falcon-40b and try merging it's lora or running live. My internet speed isn't great.

Has anyone had success applying a lora to falcon either at runtime or as a merge?
